### DISCUSSION TITLE
State management

### DISCUSSION THREAD

**Discussion by aphexmunky** [ava-labs/hypersdk#356] [Q&A]
I see the state is stored and managed from the actions. There is also state created in the controller. Am i right in thinking that this is transient and potentially each node could return different versions of the order book or is the accepted code guaranteed to execute for each transaction?

https://github.com/ava-labs/hypersdk/blob/95942cc06551fc43133060b6cda528d271240db7/examples/tokenvm/controller/controller.go#L191-L194

**Comment by patrick-ogrady** [comment]
The orders are stored on-chain, so each node should return the same version of the order book. The version in the controller is a memory-based data structure to make it efficient to serve over RPC (which is only updated on Accept).

It is important to note that the controller does not have the ability to write to "state" but rather keeps a separate DB for storing metadata (like tx hashes).
